Samples/BackgroundBlur/BackgroundBlur/SegmentModel.cpp

```diff
@@ -0,0 +1,393 @@
+}
+
+void SegmentModel::Run(const BYTE** pSrc, BYTE** pDest, DWORD cbImageSize)
+{
+	// TODO: Make all these sessions earlier, so just passing inputs/outputs here
+
+	// 1. Get input buffer as a Tensor
+	winrt::array_view<const byte> source{ *pSrc, *pSrc + cbImageSize }; // TODO: Does this work when topdown vs. bottomup
+	std::vector<int64_t> shape = { 1, 3, m_imageHeightInPixels, m_imageWidthInPixels };
+	ITensor inputRawTensor = TensorUInt8Bit::CreateFromArray(std::vector<int64_t>{1, cbImageSize}, source);
+	ITensor tensorizedImg = TensorFloat::Create(shape);
+	auto tensorizationBinding = Evaluate(m_bufferSess, std::vector<ITensor*>{&inputRawTensor}, &tensorizedImg);
+
+	// 2. Normalize input tensor
+	std::vector<float> mean = { 0.485f, 0.456f, 0.406f };
+	std::vector<float> std = { 0.229f, 0.224f, 0.225f };
+	LearningModelSession normalizationSession = CreateLearningModelSession(Normalize0_1ThenZScore(m_imageHeightInPixels, m_imageWidthInPixels, 3, mean, std));
+	ITensor intermediateTensor = TensorFloat::Create(shape);
+	auto normalizationBinding = Evaluate(normalizationSession, std::vector<ITensor*>{&tensorizedImg}, & intermediateTensor);
+
+	// 3. Run through actual model
+	std::vector<int64_t> FCNResnetOutputShape = { 1, 21, m_imageHeightInPixels, m_imageWidthInPixels };
+	LearningModelSession FCNResnetSession = CreateLearningModelSession(FCNResnet());
+	ITensor FCNResnetOutput = TensorFloat::Create(FCNResnetOutputShape);
+	auto FCNResnetBinding = Evaluate(FCNResnetSession, std::vector<ITensor*>{&intermediateTensor}, & FCNResnetOutput);
+
+	// 4.Extract labels with argmax
+	ITensor rawLabels = TensorFloat::Create({1, 1, m_imageHeightInPixels, m_imageWidthInPixels});
+	LearningModelSession argmaxSession = CreateLearningModelSession(Argmax(1, m_imageHeightInPixels, m_imageWidthInPixels));
+	auto argmaxBinding = Evaluate(argmaxSession, std::vector<ITensor*>{&FCNResnetOutput}, & rawLabels);
+
+	// 5. Get the foreground
+	ITensor foreground = TensorUInt8Bit::Create(std::vector<int64_t>{1, m_imageHeightInPixels, m_imageWidthInPixels, 3});
+	LearningModelSession foregroundSession = CreateLearningModelSession(GetBackground(1, 3, m_imageHeightInPixels, m_imageWidthInPixels));
+	// Enable tensorcpusync for the last evaluate so can extract and give back to buffer
+	// Will remove once can just pass along d3d reources back to MFT
+	auto foregroundBinding = Evaluate(foregroundSession, std::vector<ITensor*>{&tensorizedImg, &rawLabels}, & foreground, true);
+	
+	UINT32 outCapacity = 0;
+	if (m_useGPU)
+	{
+		// v1: just get the reference- should fail
+		auto reference = foreground.as<TensorUInt8Bit>().CreateReference().data();
+
+		// v3: get from a d3dresource
+		/*ID3D12Resource* res = NULL;
+		HRESULT hr = foreground.as<ITensorNative>()->GetD3D12Resource(&res);
+		UINT DstRowPitch = 0, DstDepthPitch = 0, SrcSubresource = 0;
+		hr = res->ReadFromSubresource((void*)*pDest, DstRowPitch, DstDepthPitch, SrcSubresource, NULL);*/
+		return;
```

Comment:
Is this to be done?

Reply:
yes, in the future this will be allocating d3d12 resources that the video renderer can use directly.